homeassistant/components/smartthings/climate.py

```diff
@@ -0,0 +1,226 @@
+    async_add_entities(
+        [SmartThingsThermostat(device) for device in broker.devices.values()
+         if is_climate(device)], True)
+
+
+def is_climate(device):
+    """Determine if the device should be represented as a climate entity."""
+    from pysmartthings import Capability
+
+    # Can have this legacy/deprecated capability
+    if Capability.thermostat in device.capabilities:
+        return True
+    # Or must have all of these
+    climate_capabilities = [
+        Capability.temperature_measurement,
+        Capability.thermostat_cooling_setpoint,
+        Capability.thermostat_heating_setpoint,
+        Capability.thermostat_mode]
+    if all(capability in device.capabilities
+           for capability in climate_capabilities):
+        return True
+    # Optional capabilities:
+    # relative_humidity_measurement -> state attribs
+    # thermostat_operating_state -> state attribs
+    # thermostat_fan_mode -> SUPPORT_FAN_MODE
+    return False
+
+
+class SmartThingsThermostat(SmartThingsEntity, ClimateDevice):
+    """Define a SmartThings climate entities."""
+
+    def __init__(self, device):
+        """Init the class."""
+        super().__init__(device)
+        self._base_supported_features = self._determine_base_features()
+        self._supported_features = 0
+
+    def _determine_base_features(self):
+        from pysmartthings import Capability
+
+        flags = SUPPORT_OPERATION_MODE
+        if self._device.get_capability(
+                Capability.thermostat_fan_mode, Capability.thermostat):
+            flags |= SUPPORT_FAN_MODE
+        return flags
+
+    async def async_set_fan_mode(self, fan_mode):
+        """Set new target fan mode."""
+        await self._device.set_thermostat_fan_mode(fan_mode, set_status=True)
+
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_set_operation_mode(self, operation_mode):
+        """Set new target operation mode."""
+        mode = STATE_TO_MODE[operation_mode]
+        await self._device.set_thermostat_mode(mode, set_status=True)
+
+        # State is set optimistically in the command above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_set_temperature(self, **kwargs):
+        """Set new operation mode and target temperatures."""
+        # Operation state
+        operation_state = kwargs.get(ATTR_OPERATION_MODE)
+        if operation_state:
+            mode = STATE_TO_MODE[operation_state]
+            await self._device.set_thermostat_mode(mode, set_status=True)
+
+        # Heat/cool setpoint
+        heating_setpoint = None
+        cooling_setpoint = None
+        if self.current_operation == STATE_HEAT:
```

Comment:
It's not optimal that we depend on current operation mode to determine what to do with temperature arguments. It would be better if the device presented the available options, and let those determine the logic here. But I'm guessing that's not possible?

Reply:
Fully agree, but due to the quirks of the entity this is necessary.

Reply:
Don't blame this on the entity. Many devices support setting temps regardless of operating mode, and instead handles what to do with the temperature in the device or device library.

It would be best if it doesn't matter if we set the temperature before the operating mode and vice versa. Here we're shifting the responsibility for correct operation on the user.

Reply:
Except it isn’t a library or device concern...  the lib isn’t a HAAS adapter—the entity is.  It arises because target temp can be passed individually which has to be mapped to high or low based on the current mode.  We set the operation first because if they passed in a single target for on/cool it’s obviously meant for those set points.  This actually makes it more user friendly, IMO.

Reply:
It would be more flexible if the library would handle each temperature set, single temp, low temp and high temp, and use them appropriately depending on operation mode. But if it doesn't do that we have to adapt. I'm just asking to make sure that's the case.

Reply:
Yeah -- it doesn't today.  The library represents devices as an _unrelated_ set of capabilities, attributes, and commands--which is how ST sees things.  It's up to the Entity implementation to string those together into something coherent, which means in some cases we have to emulate things to for functional reasons.  I could update the library to contain this logic instead, including the `is_climate` logic, but would need to introduce device-specific types.  Given this is opinionated (largely tied to HASS), I have to think about whether that really makes sense there, or perhaps in a different library package.

Reply:
Logic was equivalent to the wink implementation with the exception wink doesn't implement setting the operation mode.